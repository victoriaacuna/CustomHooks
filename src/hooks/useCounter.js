import React, { useState } from 'react'

export const useCounter = (initialState=10) => {

    const [state, setstate] = useState(initialState);
    const increment = () =>  {
        setstate(state+1);
    }
    const decrement = () =>  {
        setstate(state-1);
    }

    const reset = () =>  {
        setstate(initialState);
    }

    return {
        state,
        increment,
        decrement,
        reset
    }
}
