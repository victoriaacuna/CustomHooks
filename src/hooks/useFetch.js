import { useEffect, useRef, useState } from "react"


export const useFetch = (url) => {

    const isMounted = useRef(true)

    const [state, setstate] = useState({
        data: null,
        loading: true,
        error: null,
    });

    useEffect(() => {
        
        return () => {
            isMounted.current=false;
        }
    }, []);


    useEffect(() => {

        setstate({
            data: null,
            loading:true,
            error:null
        })

        fetch(url).then( resp => resp.json()).then( data => {
            setTimeout( () => {
                // Si estÃ¡ montado, hago el cambio.
                if(isMounted.current){
                    setstate({
                        loading: false,
                        error: null,
                        data
                    });
                }
            }, 1000);
            // if(isMounted.current){
            //     setstate({
            //         loading: false,
            //         error: null,
            //         data
            //     })
            // }
            
            
        }).catch( () => {
            setstate({
                loading: false,
                error: 'No se pudo cargar la info.',
                data: null
            })
        })
    }, [url]);

    return state;
    
}
