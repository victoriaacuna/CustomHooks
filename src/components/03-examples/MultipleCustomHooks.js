import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import {useCounter} from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const {state:counter, increment} = useCounter(1);

    const {loading, data} = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
   
    // Si existe la data, la constante toma ese valor.
    const {author, quote} = !!data && data[0];

    return (
        <div>
            <h1>Custom Hooks!!</h1>
            <hr/>

            {
                loading 
                ?
                (
                    <div className="alert alert-info text-center">
                        Loading...
                    </div>
                )
                :
                <>
                    <blockquote className="blockquote text-right">
                        <p className="mb-0">{quote}</p>
                        <footer className="blockquote-footer">{author}</footer>
                    </blockquote>
                    <button className="btn btn-primary" onClick={increment}>
                        Siguiente quote
                    </button>
                </>
            }
        </div>
    )
}
