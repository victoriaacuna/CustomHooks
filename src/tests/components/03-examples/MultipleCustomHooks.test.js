import React from 'react';
import {shallow} from 'enzyme';
import '@testing-library/jest-dom';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en Multiple Custom Hooks', () => {

    // Es como una falsa implementación. Finge que lo implementa y puedes usar los valores que retornó (tú especificas lo que retorna)
    useCounter.mockReturnValue({
        counter: 15,
        increment: () => {}
    })

    test('should match snapshot', () => {

        // Tienes que especificar esto porque el mock ya la prueba lo está leyendo y sino te dará error. 
        // (A pesar de que no te interese el mock acá sino en la otra prueba)
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should show info', () => {

        // Usas el mock porque no te importa la implementación, solo el resultado (el useFetch ya lo probaste)
        useFetch.mockReturnValue({
            data: [{
                author: 'Vics',
                quote: 'Hola mundo, how r u'
            }],
            loading: false,
            error: null
        });
        const wrapper = shallow(<MultipleCustomHooks/>);
        // console.log(wrapper.html());
        expect(wrapper.find('.alert').exists()).toBeFalsy();
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola mundo, how r u');
        expect(wrapper.find('footer').text().trim()).toBe('Vics');

    })
    
    
})
