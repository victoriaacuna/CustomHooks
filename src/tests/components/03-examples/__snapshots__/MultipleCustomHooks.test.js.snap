// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Multiple Custom Hooks should match snapshot 1`] = `
<div>
  <h1>
    Custom Hooks!!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</div>
`;
